ocfweb/api/session_tracking.py

```diff
@@ -0,0 +1,111 @@
+import json
+from functools import partial
+from ipaddress import ip_address
+
+from django.conf import settings
+from django.http import HttpResponse
+from django.http import HttpResponseBadRequest
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.http import require_POST
+from ocflib.infra.ldap import ldap_ocf
+from ocflib.infra.ldap import OCF_LDAP_HOSTS
+from ocflib.infra.net import is_ocf_ip
+from ocflib.lab.stats import get_connection
+
+from ocfweb.caching import cache
+
+get_connection = partial(
+    get_connection,
+    user=settings.OCFSTATS_USER,
+    password=settings.OCFSTATS_PASSWORD,
+    db=settings.OCFSTATS_DB,
+)
+
+
+@require_POST
+@csrf_exempt
+def log_session(request):
+    """Primary API endpoint for session tracking.
+
+    Desktops have a cronjob that calls this endpoint, replacing
+    the functionality that used to be in ocf/labstats.
+    """
+
+    if not is_ocf_ip(ip_address(request.META['REMOTE_ADDR'])):
+        return HttpResponse('Not Authorized', status=401)
+
+    try:
+        body = json.loads(request.body.decode('utf-8'))
+
+        host = _get_desktops().get(request.META['REMOTE_ADDR'], None)
+        user = body.get('user', None)
+
+        if user is '' or user is None:
+            raise ValueError('Invalid user "{}"'.format(user))
+
+        if host is None:
+            raise ValueError('Invalid host "{}"'.format(host))
+
+        if _session_exists(host, user):
+            _refresh_session(host, user)
+        else:
+            _new_session(host, user)
```

Comment:
is there any way to upsert this in raw sql? this is sort of yucky and race-prone as written, and uses multiple txns

Reply:
The closest thing I can think of would be like `INSERT ... ON DUPLICATE KEY UPDATE` but that doesn't solve this problem, I don't think. Perhaps a solution closer to https://github.com/ocf/ocfweb/blob/0c284bb5759da461c49b60118bc34c57cf581126/ocfweb/account/vhost_mail.py#L381-L391 like having each function use the same connection and committing the transaction at the end of the function?

Reply:
okay. sad we can't just do it in sql, but that's okay. the txn approach is better but not critical (ok with shipping this without it, it's not really a "critical" database where races matter much)